intel_adsp: ace: secondary core power-up

tmleman opened the pull request:
Changes in the sequence of entering and exiting D3 state by the secondary cores.

tmleman committed:
intel_adsp: ace: d3 exit update

This patch is moving common power configuration code outside of the
section only for the primary core. This should be enabled for all cores
and it was put there probably by mistake.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: ace: add core power off step

Adding additional core power-off before core is properly power-up after
power domains is wake up from power gaiting state.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

softwarecki commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -220,7 +220,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;
 		DSPCS.bootctl[cpu].wdtcs = DSPBR_WDT_RESTART_COMMAND;
+		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPPG;
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
```
DSPCS.bootctl[cpu].bctl &= ~(DSPBR_BCTL_WAITIPPG | DSPBR_BCTL_WAITIPCG);

softwarecki commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -96,6 +96,15 @@ void soc_start_core(int cpu_num)
 
+	/* Disabling power before core power up. This step is required if core is starting after
+	 * entering power gating and being turn off.
+	 */
```
shouldn't there be a "Disabling power **gating** before ..."

ceolin replied:
yep, this comment is a little bit hard to understand

tmleman replied:
This bit is not responsible for power gating. The comment has been updated.

ceolin commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -96,6 +96,15 @@ void soc_start_core(int cpu_num)
+	 */
+	DSPCS.capctl[cpu_num].ctl &= ~DSPCS_CTL_SPA;
+
+	while (((DSPCS.capctl[cpu_num].ctl & DSPCS_CTL_CPA) == DSPCS_CTL_CPA)) {
```
It is worth some comment here about what you are checking. These register names are not easy to udnerstand

ceolin commented:
The logic in this pr looks fine for me, please take a look in the comments

marcinszkudlinski commented:
looks fine

tmleman committed:
intel_adsp: ace: enable core power gating

Allowing the power domain of non-active cores to enter power gating
state.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: ace: enable interrupts for secondary core

Temporary re-enabling interrupts before going to waiti. Right now
secondary cores don't have proper context restore flow and after leaving
D3 state core will return here and stuck. This is temporary workaround.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

softwarecki commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -170,6 +181,7 @@ int soc_adsp_halt_cpu(int id)
 		return -EINVAL;
 	}
 
+	ACE_PWRCTL->wpdsphpxpg &= ~BIT(id);
```
What is wpdsphpxpg?

tmleman replied:
Commit message is describing what this bitfield does. In the future, it should be done with usage of power_domain_intel_adsp.

softwarecki replied:
When you read a file and try to understand what it does, you don't do git blame to read all the commit messages. The power control code is full of names that look like the result of obfuscation (eg. wpdsphpxpg, wphstpg, wphubhppg, wpdspulppg, wpioxpg, wpmlpg, phubulppg, dsphpxpgs, hstpgs, hubhppgs, dspulppgs, ioxpgs, mlpgs, hubulppgs). We should definitely add comments with human names, add one-sentence descriptions of what we're doing when toggling some bit. Without it, it's very easy to make a mistake and code maintenance becomes impossible.

aborisovich reviewed (approved):
Looks good to me

stephanosio commented:
@tmleman A bug report is required in order for this PR to be merged during the feature freeze. Please file a bug report issue for the problem that this PR is trying to fix and link the issue to this PR.

nashif commented:
@stephanosio  bug added
